File: src/praisonai-agents/praisonaiagents/runtime/registry.py
"""Runtime registry for PraisonAI Agents.

This module provides both protocol-based registry interface (from main) and
a concrete implementation (from this branch) for runtime registration and resolution.

Protocol-driven design following AGENTS.md:
- Core protocols only (no heavy implementations at module level)
- Thread-safe registration with factory pattern
- Built-in runtime support
- Entry points discovery for plugin runtimes
- Fail-closed behavior for unknown runtime IDs
"""

import threading
from typing import Protocol, Dict, Callable, Any, Optional, List, runtime_checkable
from dataclasses import dataclass
from .protocols import AgentRuntimeProtocol


# Registry entry dataclass from main
@dataclass
class RuntimeRegistryEntry:
    """Entry in the runtime registry containing runtime metadata."""
    runtime_id: str
    display_name: Optional[str] = None
    description: Optional[str] = None
    is_builtin: bool = False
    metadata: Dict[str, Any] = None
    
    def __post_init__(self):
        if self.metadata is None:
            self.metadata = {}
        if self.display_name is None:
            self.display_name = self.runtime_id


# Protocol definition from main
@runtime_checkable
class RuntimeRegistryProtocol(Protocol):
    """Protocol for runtime registries.
    
    Defines the interface for registering and resolving agent runtimes.
    The actual implementation lives in the wrapper layer.
    """
    
    def register(self, runtime_id: str, factory_or_instance: Any, **metadata) -> None:
        """Register a runtime with the registry.
        
        Args:
            runtime_id: Unique runtime identifier (e.g., "claude-code")
            factory_or_instance: Runtime factory function or instance
            **metadata: Additional metadata for the runtime
        
        Raises:
            ValueError: If runtime_id is already registered
            TypeError: If factory_or_instance is not callable or valid instance
        """
        ...
    
    def unregister(self, runtime_id: str) -> bool:
        """Unregister a runtime from the registry.
        
        Args:
            runtime_id: Runtime identifier to unregister
            
        Returns:
            True if runtime was unregistered, False if not found
        """
        ...
    
    def is_registered(self, runtime_id: str) -> bool:
        """Check if a runtime is registered.
        
        Args:
            runtime_id: Runtime identifier to check
            
        Returns:
            True if runtime is registered, False otherwise
        """
        ...
    
    def list_runtimes(self) -> List[RuntimeRegistryEntry]:
        """List all registered runtimes.
        
        Returns:
            List of RuntimeRegistryEntry objects for all registered runtimes
        """
        ...
    
    def get_entry(self, runtime_id: str) -> Optional[RuntimeRegistryEntry]:
        """Get registry entry for a runtime.
        
        Args:
            runtime_id: Runtime identifier
            
        Returns:
            RuntimeRegistryEntry if found, None otherwise
        """
        ...
    
    def resolve(self, runtime_id: str, config_overrides: Optional[Dict[str, Any]] = None) -> Any:
        """Resolve a runtime by ID with optional configuration overrides.
        
        Args:
            runtime_id: Runtime identifier to resolve
            config_overrides: Optional configuration overrides
            
        Returns:
            Runtime instance (typically implementing AgentRuntimeProtocol)
            
        Raises:
            ValueError: If runtime_id is not registered (fail-closed behavior)
            TypeError: If runtime cannot be instantiated
        """
        ...
    
    def clear(self) -> None:
        """Clear all registered runtimes.
        
        Used primarily for testing and cleanup.
        """
        ...


# Global registry state from this branch
_runtimes: Dict[str, Callable[[], AgentRuntimeProtocol]] = {}
_runtime_aliases: Dict[str, str] = {}
_runtime_metadata: Dict[str, Dict[str, Any]] = {}
_registry_lock = threading.RLock()
_builtin_initialized = False


def _get_builtin_runtime_factories() -> Dict[str, Callable[[], Callable[[], AgentRuntimeProtocol]]]:
    """Get built-in runtime factory loaders."""
    def praisonai_loader():
        def factory():
            from .builtin import PraisonAIRuntime
            return PraisonAIRuntime()
        return factory
    
    return {
        "praisonai": praisonai_loader
    }


def _initialize_builtin_runtimes() -> None:
    """Initialize built-in runtimes if not already done."""
    global _builtin_initialized
    
    if _builtin_initialized:
        return
        
    builtin_factories = _get_builtin_runtime_factories()
    
    with _registry_lock:
        if not _builtin_initialized:
            for runtime_id, loader in builtin_factories.items():
                if runtime_id not in _runtimes:
                    _runtimes[runtime_id] = loader()
                    _runtime_metadata[runtime_id] = {"is_builtin": True}
            _builtin_initialized = True


# Concrete implementation that implements the protocol
class RuntimeRegistry:
    """Concrete runtime registry implementation.
    
    This provides both the RuntimeRegistryProtocol interface and additional
    helper methods for managing runtimes.
    """
    
    def register(self, runtime_id: str, factory_or_instance: Any, **metadata) -> None:
        """Register a runtime with the registry."""
        if not callable(factory_or_instance):
            # If it's an instance, wrap in a factory
            def factory():
                return factory_or_instance
            factory_func = factory
        else:
            factory_func = factory_or_instance
        
        with _registry_lock:
            if runtime_id in _runtimes:
                raise ValueError(f"Runtime '{runtime_id}' is already registered")
            _runtimes[runtime_id] = factory_func
            _runtime_metadata[runtime_id] = metadata
    
    def unregister(self, runtime_id: str) -> bool:
        """Unregister a runtime."""
        with _registry_lock:
            if runtime_id in _runtimes:
                del _runtimes[runtime_id]
                if runtime_id in _runtime_metadata:
                    del _runtime_metadata[runtime_id]
                # Remove any aliases pointing to this runtime
                aliases_to_remove = [
                    alias for alias, canonical in _runtime_aliases.items()
                    if canonical == runtime_id
                ]
                for alias in aliases_to_remove:
                    del _runtime_aliases[alias]
                return True
            return False
    
    def is_registered(self, runtime_id: str) -> bool:
        """Check if a runtime is registered."""
        _initialize_builtin_runtimes()
        with _registry_lock:
            canonical_id = _runtime_aliases.get(runtime_id, runtime_id)
            return canonical_id in _runtimes
    
    def list_names(self) -> List[str]:
        """List all registered runtime names (IDs).
        
        Returns:
            List of runtime IDs
        """
        _initialize_builtin_runtimes()
        with _registry_lock:
            return list(_runtimes.keys())
    
    def is_available(self, runtime_id: str) -> bool:
        """Check if a runtime is available (registered).
        
        This is an alias for is_registered() to match the test expectations.
        
        Args:
            runtime_id: Runtime identifier
            
        Returns:
            True if runtime is registered or aliased, False otherwise
        """
        return self.is_registered(runtime_id)
    
    def list_runtimes(self) -> List[RuntimeRegistryEntry]:
        """List all registered runtimes."""
        _initialize_builtin_runtimes()
        with _registry_lock:
            entries = []
            for runtime_id in _runtimes:
                metadata = _runtime_metadata.get(runtime_id, {})
                entries.append(RuntimeRegistryEntry(
                    runtime_id=runtime_id,
                    display_name=metadata.get("display_name"),
                    description=metadata.get("description"),
                    is_builtin=metadata.get("is_builtin", False),
                    metadata=metadata
                ))
            return entries
    
    def list_names(self) -> List[str]:
        """List all registered runtime names/IDs."""
        entries = self.list_runtimes()
        return [e.runtime_id for e in entries]
    
    def get_entry(self, runtime_id: str) -> Optional[RuntimeRegistryEntry]:
        """Get registry entry for a runtime."""
        _initialize_builtin_runtimes()
        with _registry_lock:
            canonical_id = _runtime_aliases.get(runtime_id, runtime_id)
            if canonical_id not in _runtimes:
                return None
            metadata = _runtime_metadata.get(canonical_id, {})
            return RuntimeRegistryEntry(
                runtime_id=canonical_id,
                display_name=metadata.get("display_name"),
                description=metadata.get("description"),
                is_builtin=metadata.get("is_builtin", False),
                metadata=metadata
            )
    
    def resolve(self, runtime_id: str, config_overrides: Optional[Dict[str, Any]] = None) -> Any:
        """Resolve a runtime by ID."""
        _initialize_builtin_runtimes()
        
        with _registry_lock:
            # Check for alias
            canonical_id = _runtime_aliases.get(runtime_id, runtime_id)
            
            if canonical_id not in _runtimes:
                available = sorted(_runtimes.keys())
                raise ValueError(f"Unknown runtime: {runtime_id}. Available: {available}")
            
            factory = _runtimes[canonical_id]
        
        # Create instance outside of lock to prevent deadlock
        return factory()
    
    def clear(self) -> None:
        """Clear all registered runtimes."""
        global _builtin_initialized
        with _registry_lock:
            _runtimes.clear()
            _runtime_aliases.clear()
            _runtime_metadata.clear()
            _builtin_initialized = False
    
    def add_alias(self, alias: str, canonical_runtime_id: str) -> None:
        """Add an alias for a runtime."""
        _initialize_builtin_runtimes()
        
        with _registry_lock:
            if canonical_runtime_id not in _runtimes:
                available = sorted(_runtimes.keys())
                raise ValueError(f"Cannot create alias '{alias}' for unknown runtime: {canonical_runtime_id}. Available: {available}")
            
            _runtime_aliases[alias] = canonical_runtime_id
    
    def is_available(self, runtime_id: str) -> bool:
        """Check if a runtime is available."""
        try:
            self.resolve(runtime_id)
            return True
        except ValueError:
            return False


# Global registry instance
_global_registry = RuntimeRegistry()


# Convenience functions that use the global registry
def register_runtime(runtime_id: str, factory: Callable[[], AgentRuntimeProtocol], **metadata) -> None:
    """Register a runtime factory with the global registry.
    
    Args:
        runtime_id: Unique identifier (e.g., "praisonai", "claude-code") 
        factory: Factory function that returns an AgentRuntimeProtocol instance
        **metadata: Additional metadata for the runtime
    """
    _global_registry.register(runtime_id, factory, **metadata)


def unregister_runtime(runtime_id: str) -> bool:
    """Unregister a runtime from the global registry.
    
    Args:
        runtime_id: Runtime identifier
        
    Returns:
        True if runtime was found and removed, False otherwise
    """
    return _global_registry.unregister(runtime_id)


def list_runtimes() -> list[str]:
    """List all registered runtime IDs."""
    entries = _global_registry.list_runtimes()
    return [e.runtime_id for e in entries]


def resolve_runtime(runtime_id: str) -> AgentRuntimeProtocol:
    """Resolve a runtime by ID using the global registry.
    
    Args:
        runtime_id: Runtime identifier (e.g., "praisonai")
        
    Returns:
        AgentRuntimeProtocol instance
        
    Raises:
        ValueError: If runtime_id is not registered
    """
    return _global_registry.resolve(runtime_id)


def add_runtime_alias(alias: str, canonical_runtime_id: str) -> None:
    """Add an alias for a runtime.
    
    Args:
        alias: Alias name
        canonical_runtime_id: The canonical runtime ID this alias points to
        
    Raises:
        ValueError: If canonical runtime ID is not registered
    """
    _global_registry.add_alias(alias, canonical_runtime_id)


def is_runtime_available(runtime_id: str) -> bool:
    """Check if a runtime is available.
    
    Args:
        runtime_id: Runtime identifier
        
    Returns:
        True if runtime exists and can be created
    """
    try:
        resolve_runtime(runtime_id)
        return True
    except ValueError:
        return False


# Auto-discovery of entry points
def _discover_entry_point_runtimes() -> None:
    """Discover runtimes from entry points."""
    try:
        from importlib.metadata import entry_points
        
        for ep in entry_points(group="praisonai.runtimes"):
            def make_loader(entry_point):
                def loader():
                    factory_class = entry_point.load()
                    return factory_class()
                return loader
            
            with _registry_lock:
                if ep.name not in _runtimes:
                    _runtimes[ep.name] = make_loader(ep)
                    _runtime_metadata[ep.name] = {"entry_point": str(ep)}
                    
    except ImportError:
        # entry_points not available in older Python versions
        pass
    except Exception:
        # Ignore discovery errors - plugins are optional
        pass


# Initialize on module load
_discover_entry_point_runtimes()

File: src/praisonai-ts/src/runtime/registry.ts
/**
 * Runtime registry for PraisonAI agents.
 *
 * Port of `praisonaiagents/runtime/registry.py`: a module-level registry of
 * runtime factories with aliases, metadata, a lazily-registered `praisonai`
 * builtin, and fail-closed resolution of unknown ids.
 *
 * Behaviour mirrored from Python:
 * - `register` throws on a duplicate id (Python `ValueError`).
 * - `register` does NOT initialise the builtins first, so a caller may
 *   register its own `"praisonai"` before any lookup; the later builtin
 *   initialisation skips ids that already exist.
 * - `resolve` / `addAlias` on an unknown id throw
 *   `Unknown runtime: <id>. Available: ['a', 'b']` (same text as Python).
 * - `isAvailable` swallows only registry errors (Python catches `ValueError`
 *   only); an exception thrown by the factory itself propagates.
 * - `resolve` accepts `configOverrides` and — exactly like Python — does not
 *   forward it to the factory; it is kept for signature parity.
 * - Python defines `list_names` and `is_available` twice on the class; the
 *   later definitions win and are the ones ported.
 *
 * Not ported: entry-point discovery (`_discover_entry_point_runtimes`, the
 * `praisonai.runtimes` importlib group). Node has no equivalent of Python
 * entry points; plugin runtimes register explicitly via
 * {@link registerRuntime}.
 */

import {
  AgentRuntimeProtocol,
  RuntimeCapabilityMatrix,
  RuntimeDelta,
  RuntimeResult,
  RunTurnOptions,
} from './protocols';

/** Constructor options for {@link RuntimeRegistryEntry}. */
export interface RuntimeRegistryEntryInit {
  runtimeId: string;
  /** Default: `runtimeId`. */
  displayName?: string | null;
  /** Default `null`. */
  description?: string | null;
  /** Default `false`. */
  isBuiltin?: boolean;
  /** Default `{}`. */
  metadata?: Record<string, unknown> | null;
}

/** Entry in the runtime registry (Python `registry.py::RuntimeRegistryEntry`). */
export class RuntimeRegistryEntry {
  runtimeId: string;
  displayName: string;
  description: string | null;
  isBuiltin: boolean;
  metadata: Record<string, unknown>;

  constructor(config: RuntimeRegistryEntryInit) {
    this.runtimeId = config.runtimeId;
    this.displayName = config.displayName ?? config.runtimeId;
    this.description = config.description ?? null;
    this.isBuiltin = config.isBuiltin ?? false;
    this.metadata = config.metadata ?? {};
  }
}

/** A factory producing a runtime instance (Python `Callable[[], AgentRuntimeProtocol]`). */
export type RuntimeFactory = () => AgentRuntimeProtocol;

/**
 * Error raised for invalid registry operations — duplicate registration,
 * unknown runtime id, alias to an unknown runtime. Python raises
 * `ValueError` for all three.
 */
export class RuntimeRegistryError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'RuntimeRegistryError';
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

/** Protocol for runtime registries (Python `registry.py::RuntimeRegistryProtocol`). */
export interface RuntimeRegistryProtocol {
  register(
    runtimeId: string,
    factoryOrInstance: RuntimeFactory | AgentRuntimeProtocol,
    metadata?: Record<string, unknown>,
  ): void;
  unregister(runtimeId: string): boolean;
  isRegistered(runtimeId: string): boolean;
  listRuntimes(): RuntimeRegistryEntry[];
  getEntry(runtimeId: string): RuntimeRegistryEntry | null;
  resolve(runtimeId: string, configOverrides?: Record<string, unknown> | null): AgentRuntimeProtocol;
  clear(): void;
}

// ---------------------------------------------------------------------------
// Module-level registry state (Python `_runtimes`, `_runtime_aliases`, ...)
// ---------------------------------------------------------------------------

const _runtimes = new Map<string, RuntimeFactory>();
const _runtimeAliases = new Map<string, string>();
const _runtimeMetadata = new Map<string, Record<string, unknown>>();
let _builtinInitialized = false;

/** Python list repr for the "Available: [...]" part of error messages. */
function reprList(items: string[]): string {
  return `[${items.map((item) => `'${item}'`).join(', ')}]`;
}

/** Built-in runtime factory loaders (Python `_get_builtin_runtime_factories`). */
function _getBuiltinRuntimeFactories(): Record<string, () => RuntimeFactory> {
  return {
    praisonai: () => () => new PraisonAIRuntime(),
  };
}

/** Initialise built-in runtimes once (Python `_initialize_builtin_runtimes`). */
function _initializeBuiltinRuntimes(): void {
  if (_builtinInitialized) return;
  const builtinFactories = _getBuiltinRuntimeFactories();
  for (const [runtimeId, loader] of Object.entries(builtinFactories)) {
    if (!_runtimes.has(runtimeId)) {
      _runtimes.set(runtimeId, loader());
      _runtimeMetadata.set(runtimeId, { is_builtin: true });
    }
  }
  _builtinInitialized = true;
}

/**
 * Concrete runtime registry implementation (Python `registry.py::RuntimeRegistry`).
 *
 * All instances share the module-level state, exactly as in Python.
 */
export class RuntimeRegistry implements RuntimeRegistryProtocol {
  /**
   * Register a runtime (Python `register`). A non-function value is treated
   * as an instance and wrapped in a factory that always returns it.
   *
   * @throws RuntimeRegistryError if `runtimeId` is already registered.
   */
  register(
    runtimeId: string,
    factoryOrInstance: RuntimeFactory | AgentRuntimeProtocol,
    metadata: Record<string, unknown> = {},
  ): void {
    const factoryFunc: RuntimeFactory =
      typeof factoryOrInstance === 'function'
        ? (factoryOrInstance as RuntimeFactory)
        : () => factoryOrInstance as AgentRuntimeProtocol;

    if (_runtimes.has(runtimeId)) {
      throw new RuntimeRegistryError(`Runtime '${runtimeId}' is already registered`);
    }
    _runtimes.set(runtimeId, factoryFunc);
    _runtimeMetadata.set(runtimeId, metadata);
  }

  /** Unregister a runtime and any aliases pointing to it (Python `unregister`). */
  unregister(runtimeId: string): boolean {
    if (!_runtimes.has(runtimeId)) return false;
    _runtimes.delete(runtimeId);
    _runtimeMetadata.delete(runtimeId);
    for (const [alias, canonical] of Array.from(_runtimeAliases.entries())) {
      if (canonical === runtimeId) _runtimeAliases.delete(alias);
    }
    return true;
  }

  /** True if `runtimeId` (or an alias of it) is registered (Python `is_registered`). */
  isRegistered(runtimeId: string): boolean {
    _initializeBuiltinRuntimes();
    const canonicalId = _runtimeAliases.get(runtimeId) ?? runtimeId;
    return _runtimes.has(canonicalId);
  }

  /** All registered runtime ids (Python `list_names`, later definition). */
  listNames(): string[] {
    return this.listRuntimes().map((entry) => entry.runtimeId);
  }

  /**
   * True if the runtime resolves (Python `is_available`, later definition:
   * tries `resolve` and swallows only `ValueError`).
   */
  isAvailable(runtimeId: string): boolean {
    try {
      this.resolve(runtimeId);
      return true;
    } catch (err) {
      if (err instanceof RuntimeRegistryError) return false;
      throw err;
    }
  }

  /** All registered runtimes as entries (Python `list_runtimes`). */
  listRuntimes(): RuntimeRegistryEntry[] {
    _initializeBuiltinRuntimes();
    const entries: RuntimeRegistryEntry[] = [];
    for (const runtimeId of _runtimes.keys()) {
      const metadata = _runtimeMetadata.get(runtimeId) ?? {};
      entries.push(entryFromMetadata(runtimeId, metadata));
    }
    return entries;
  }

  /** Registry entry for a runtime (or alias), or `null` (Python `get_entry`). */
  getEntry(runtimeId: string): RuntimeRegistryEntry | null {
    _initializeBuiltinRuntimes();
    const canonicalId = _runtimeAliases.get(runtimeId) ?? runtimeId;
    if (!_runtimes.has(canonicalId)) return null;
    return entryFromMetadata(canonicalId, _runtimeMetadata.get(canonicalId) ?? {});
  }

  /**
   * Resolve a runtime by id or alias (Python `resolve`). Fail-closed: an
   * unknown id throws {@link RuntimeRegistryError}.
   *
   * `configOverrides` is accepted for parity with Python, which also does
   * not forward it to the factory.
   */
  resolve(runtimeId: string, configOverrides: Record<string, unknown> | null = null): AgentRuntimeProtocol {
    void configOverrides;
    _initializeBuiltinRuntimes();
    const canonicalId = _runtimeAliases.get(runtimeId) ?? runtimeId;
    const factory = _runtimes.get(canonicalId);
    if (factory === undefined) {
      const available = Array.from(_runtimes.keys()).sort();
      throw new RuntimeRegistryError(`Unknown runtime: ${runtimeId}. Available: ${reprList(available)}`);
    }
    return factory();
  }

  /** Clear all runtimes, aliases and metadata; builtins re-register lazily (Python `clear`). */
  clear(): void {
    _runtimes.clear();
    _runtimeAliases.clear();
    _runtimeMetadata.clear();
    _builtinInitialized = false;
  }

  /**
   * Add an alias for a registered runtime (Python `add_alias`).
   *
   * @throws RuntimeRegistryError if `canonicalRuntimeId` is not registered.
   */
  addAlias(alias: string, canonicalRuntimeId: string): void {
    _initializeBuiltinRuntimes();
    if (!_runtimes.has(canonicalRuntimeId)) {
      const available = Array.from(_runtimes.keys()).sort();
      throw new RuntimeRegistryError(
        `Cannot create alias '${alias}' for unknown runtime: ${canonicalRuntimeId}. Available: ${reprList(available)}`,
      );
    }
    _runtimeAliases.set(alias, canonicalRuntimeId);
  }
}

function entryFromMetadata(runtimeId: string, metadata: Record<string, unknown>): RuntimeRegistryEntry {
  return new RuntimeRegistryEntry({
    runtimeId,
    displayName: (metadata.display_name as string | undefined) ?? (metadata.displayName as string | undefined) ?? null,
    description: (metadata.description as string | undefined) ?? null,
    isBuiltin: Boolean(metadata.is_builtin ?? metadata.isBuiltin ?? false),
    metadata,
  });
}

// ---------------------------------------------------------------------------
// Built-in runtime (Python `runtime/builtin.py::PraisonAIRuntime`)
// ---------------------------------------------------------------------------

/** Minimal surface of the agent the builtin runtime drives. */
export interface RuntimeAgentLike {
  chat(prompt: string): Promise<string>;
  model?: unknown;
  llm?: unknown;
  id?: unknown;
  agentId?: unknown;
}

/** Options for {@link PraisonAIRuntime} (TypeScript-only; Python `__init__` takes none). */
export interface PraisonAIRuntimeOptions {
  /**
   * Builds the agent for a turn. Defaults to `new Agent(config)` from
   * `../agent/simple`, imported lazily to avoid a circular import. Injectable
   * for tests.
   */
  agentFactory?: (config: Record<string, unknown>) => Promise<RuntimeAgentLike> | RuntimeAgentLike;
}

async function defaultAgentFactory(config: Record<string, unknown>): Promise<RuntimeAgentLike> {
  const mod = await import('../agent/simple');
  return new mod.Agent(config as never) as unknown as RuntimeAgentLike;
}

function hasEnv(name: string): boolean {
  const env = (globalThis as { process?: { env?: Record<string, string | undefined> } }).process?.env;
  return Boolean(env?.[name]);
}

/**
 * Built-in PraisonAI runtime (Python `runtime/builtin.py::PraisonAIRuntime`).
 *
 * Wraps the existing `Agent` so a runtime abstraction exists without
 * duplicating the chat loop. As in Python, `streamTurn` yields the full
 * response as a single delta (the agent has no async streaming iterator).
 *
 * Python's builtin only implements `supports`, `run_turn` and `stream_turn`;
 * the remaining protocol methods here are thin delegations so the class
 * satisfies {@link AgentRuntimeProtocol} structurally.
 */
export class PraisonAIRuntime implements AgentRuntimeProtocol {
  readonly runtimeId = 'praisonai';
  readonly runtimeName = 'praisonai';
  readonly runtimeVersion = '1.0';
  private readonly agentFactory: NonNullable<PraisonAIRuntimeOptions['agentFactory']>;

  constructor(options: PraisonAIRuntimeOptions = {}) {
    this.agentFactory = options.agentFactory ?? defaultAgentFactory;
  }

  /** Capability matrix; `runtime/capabilities.py` is not ported, so this is minimal. */
  capabilities(): RuntimeCapabilityMatrix {
    return { streaming_deltas: false };
  }

  /** The builtin runtime delegates to the LLM subsystem, so it is universal (Python `supports`). */
  supports(modelRef: string | null = null): boolean {
    void modelRef;
    return true;
  }

  private buildAgentConfig(options: Record<string, unknown>): Record<string, unknown> {
    const config: Record<string, unknown> = {};
    const modelRef = options.modelRef ?? options.model_ref;
    const systemPrompt = options.systemPrompt ?? options.system_prompt;
    if (modelRef) config.llm = modelRef;
    if (systemPrompt) config.instructions = systemPrompt;
    if ('tools' in options) config.tools = options.tools;
    return config;
  }

  /** Execute a single turn (Python `run_turn`). Errors become `RuntimeResult.error`. */
  async runTurn(prompt: string, options: RunTurnOptions = {}): Promise<RuntimeResult> {
    try {
      const agent = await this.agentFactory(this.buildAgentConfig(options));
      const result = await agent.chat(prompt);
      const metadata: Record<string, unknown> = {
        model: agent.model ?? agent.llm ?? null,
        agent_id: agent.agentId ?? agent.id ?? null,
        runtime: this.runtimeId,
      };
      if (!result && !hasEnv('OPENAI_API_KEY')) {
        return new RuntimeResult({
          content: '',
          metadata,
          error: 'OPENAI_API_KEY environment variable is required',
        });
      }
      return new RuntimeResult({ content: result ? String(result) : '', metadata });
    } catch (err) {
      return new RuntimeResult({
        content: '',
        metadata: { runtime: this.runtimeId },
        error: err instanceof Error ? err.message : String(err),
      });
    }
  }

  /** Stream the response as a single delta (Python `stream_turn` workaround). */
  async *streamTurn(prompt: string, kwargs: Record<string, unknown> = {}): AsyncIterable<RuntimeDelta> {
    const result = await this.runTurn(prompt, kwargs);
    if (result.error !== null) {
      yield new RuntimeDelta({ type: 'error', content: result.error, metadata: { runtime: this.runtimeId } });
      return;
    }
    yield new RuntimeDelta({ type: 'text', content: result.content, metadata: { runtime: this.runtimeId } });
  }

  /** TypeScript addition: run a turn from an agent config dict. */
  async executeAgent(
    agentConfig: Record<string, unknown>,
    prompt: string,
    kwargs: Record<string, unknown> = {},
  ): Promise<Record<string, unknown>> {
    const result = await this.runTurn(prompt, { ...agentConfig, ...kwargs });
    return { response: result.content, metadata: result.metadata, error: result.error };
  }

  /** TypeScript addition: stream a turn from an agent config dict. */
  streamAgent(
    agentConfig: Record<string, unknown>,
    prompt: string,
    kwargs: Record<string, unknown> = {},
  ): AsyncIterable<RuntimeDelta> {
    return this.streamTurn(prompt, { ...agentConfig, ...kwargs });
  }

  /** TypeScript addition: the builtin accepts any config. */
  async validateConfig(agentConfig: Record<string, unknown>): Promise<string[]> {
    void agentConfig;
    return [];
  }

  /** TypeScript addition: the builtin is always healthy when importable. */
  async healthCheck(): Promise<Record<string, unknown>> {
    return { status: 'ok', runtime: this.runtimeId };
  }
}

// ---------------------------------------------------------------------------
// Global registry and convenience functions
// ---------------------------------------------------------------------------

const _globalRegistry = new RuntimeRegistry();

/** The global runtime registry instance (TypeScript accessor for Python `_global_registry`). */
export function getRuntimeRegistry(): RuntimeRegistry {
  return _globalRegistry;
}

/** Register a runtime factory with the global registry (Python `register_runtime`). */
export function registerRuntime(
  runtimeId: string,
  factory: RuntimeFactory | AgentRuntimeProtocol,
  metadata: Record<string, unknown> = {},
): void {
  _globalRegistry.register(runtimeId, factory, metadata);
}

/** Unregister a runtime from the global registry (Python `unregister_runtime`). */
export function unregisterRuntime(runtimeId: string): boolean {
  return _globalRegistry.unregister(runtimeId);
}

/** All registered runtime ids (Python `list_runtimes`). */
export function listRuntimes(): string[] {
  return _globalRegistry.listRuntimes().map((entry) => entry.runtimeId);
}

/** Resolve a runtime by id using the global registry (Python `resolve_runtime`). */
export function resolveRuntime(runtimeId: string): AgentRuntimeProtocol {
  return _globalRegistry.resolve(runtimeId);
}

/** Add an alias for a runtime (Python `add_runtime_alias`). */
export function addRuntimeAlias(alias: string, canonicalRuntimeId: string): void {
  _globalRegistry.addAlias(alias, canonicalRuntimeId);
}

/** True if the runtime exists and can be created (Python `is_runtime_available`). */
export function isRuntimeAvailable(runtimeId: string): boolean {
  try {
    resolveRuntime(runtimeId);
    return true;
  } catch (err) {
    if (err instanceof RuntimeRegistryError) return false;
    throw err;
  }
}
